docs/couler-api-design.md

```diff
@@ -0,0 +1,87 @@
+# Couler API Design
+
+This document outlines the design of core Couler APIs to support multiple workflow backends.
+
+## Goals
+
+* Design the core Couler APIs that can be implemented across different workflow engines.
+* Provide a *minimal* set of APIs that are forward-looking, engine-agnostic, and less likely to deprecate over-time.
+
+## Non-Goals
+
+* Provide an exhaustive set of APIs that cover all use cases and all workflow engines.
+* Provide implementation details.
+
+## Design
+
+Core operations (`couler.ops`):
+
+* `run_step(step_def)` where `step_def` is the step definition that can be a container spec, Python function,
+or spec that's specific to the underlying workflow engine (e.g. k8s CRD if Argo Workflow is used).
+
+Control flow (`couler.control_flows`):
+
+* `map(func, *args, **kwargs)` where `*args` and `**kwargs` correspond to various arguments passed to the function `func`.
+* `when(cond, if_op, else_op)` where
+    * `cond` can be any of the following predicates: `equal()`, `not_equal()`, `bigger()`, `smaller()`, `bigger_equal()`, `smaller_equal()`.
+    * The operation defined in `if_op` will be executed when `cond` is true. Otherwise, `else_op` will be executed.
+* `while_loop(cond, func, *args, **kwargs)` where
+    * `cond` can be any one of the predicates mentioned above.
+    *  `func`, `*args`, and `**kwargs` are similar to `map()`'s.
+
+Utilities (`couler.utils`):
+
+* `submit(config=workflow_config(schedule="* * * * 1"))` where `config` is engine-specific.
+* `get_status(workflow_name)`
+* `get_logs(workflow_name)`
+* `delete_workflow(workflow_name)`
+
+Backends (`couler.backends`):
+
+* `get_backend()`
+* `use_backend("argo")`
+
```

Comment:
Can we add a parameters section? (or at least think about that behavior?)

I think the matrix definition proposed above is great, also would love to see how parameters are handled. I know that you may define global parameters, which were shared across the steps in a workflow (for airflow at least). This could be achieved in argo, I'm not sure how that is handled in tekton.

Reply:
OK I see that is a future thing (input/output/parameters), though from my perspective this is one thing we want to get to early

Reply:
We currently support input parameters via `args`, e.g. see [this test](https://github.com/couler-proj/couler/blob/master/couler/tests/input_parameter_test.py) for examples. However I have not put thoughts into other backends for parameters yet. It might require a dedicated design doc for that.